Feature/pro 3484 android

germanRodriguezGarcia opened the pull request.

germanRodriguezGarcia committed:
WIP

germanRodriguezGarcia committed:
arscenehandler listens location updates, navigation updates and receives building info

germanRodriguezGarcia committed:
instead of painting route, paint route segments

germanRodriguezGarcia committed:
Fixes. Do not redraw route on each navigation progress, unsubscribe location and navigation listeners onpause

germanRodriguezGarcia committed:
Fixed route nodes positions, added debug buttons to redraw world and route, and to update arrow target.

germanRodriguezGarcia committed:
added basic logic to update arrow target

germanRodriguezGarcia committed:
added discs with poi images

germanRodriguezGarcia committed:
wip

germanRodriguezGarcia committed:
fix

germanRodriguezGarcia committed:
Loading models for pois, wip

germanRodriguezGarcia committed:
Added ar refresh quality metric

germanRodriguezGarcia committed:
added reset threshold

germanRodriguezGarcia committed:
added debug info and refresh using basic quality estaimation algorithm

germanRodriguezGarcia committed:
wip

germanRodriguezGarcia committed:
wip

germanRodriguezGarcia committed:
fixes

germanRodriguezGarcia committed:
fix. Pois always lookat camera. On debug do not remove route nodes, only make them invisible

germanRodriguezGarcia committed:
set visual odometry to sdk

germanRodriguezGarcia committed:
debug , added button to activate / deactivate route spheres

germanRodriguezGarcia committed:
Fixes and logs. Check if already refreshing and do not regenearete always geometry

germanRodriguezGarcia committed:
logs and debug

germanRodriguezGarcia committed:
Use modified lib version with exepeption handling on destroy method. Added long press to back button in order to enable debug options.

germanRodriguezGarcia committed:
Debug performance. Commented mapviewer, added debug button to add random pois to scene. Fixed pois load function.

germanRodriguezGarcia committed:
Refactor, geometry node with poi disk and view node with textview now hang from same base node.

germanRodriguezGarcia committed:
Adding models to geofences. WIP

germanRodriguezGarcia committed:
trying to hide viewer when ar wip

germanRodriguezGarcia committed:
On destination reached return from AR. Adjusted poi materials. Set render distance to 15m. Added style with shadows to poi names. Added new direction Arrow model.

germanRodriguezGarcia committed:
clean

germanRodriguezGarcia committed:
clean

germanRodriguezGarcia committed:
Add debug info to native

germanRodriguezGarcia committed:
Remove ar debug utils from flutter

germanRodriguezGarcia committed:
remove no needed code. Position models in a cone in front of  the camera when entering a fence

germanRodriguezGarcia committed:
clean

germanRodriguezGarcia committed:
remove no needed code

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/ARController.kt:
```diff
@@ -12,33 +20,81 @@ class ARController(
     private val arSceneHandler: ARSceneHandler,
     private val arMethodCallSender: ARMethodCallSender,
-) : DefaultLifecycleObserver {
+) : DefaultLifecycleObserver, ARControllerCallback {
```
Temos unha dependencia cíclica entre ARController e ARSceneHandler, penso que por eso meteches esta interface. É boa idea, pero para min o nome sería máis ben ARSceneHandlerCallback: a interface defínese en ARSceneHandler e a acción orixínase alí; ARController implementa a interface e reacciona ante a callback.

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/ARController.kt:
```diff
@@ -12,33 +20,81 @@ class ARController(
     }
 
+    init {
+        arSceneHandler.setCallback(this)
```
Esto está ben porque non podes pasarllo no constructor, pero non deixa de ser inxección de dependencia e podes facelo millor onde fixemos toda a DI, en SitumARPlatformView.

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/ARController.kt:
```diff
@@ -12,33 +20,81 @@ class ARController(
         isLoaded = true
         isLoading = false
+
+        handler.post(updateDebugInfoRunnable)
```
As movidas de debug como esta meteríaas debaixo dunha flag, para que non afecten ao rendemento si non está en "modo debug".

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/ARController.kt:
```diff
@@ -71,4 +127,13 @@ class ARController(
+    }
+
+
+    override fun sendARGone() {
```
Si é unha callback, un nome máis intuitivo sería algo en plan "onArGoneRequired" ou "onArShouldStop", ou algo que nos leve a pensar que é unha resposta a un evento.

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/ARController.kt:
```diff
@@ -71,4 +127,13 @@ class ARController(
+
+
+    override fun sendARGone() {
+        arMethodCallSender.sendArGoneRequired()
```
Aí arriba aparece tamén unha chamada a sendArGoneRequired, protexida baixo unha flag. Igual che vale a pena crear método privado e invocar o mismo desde ambos sitios, ou aproveitar a misma implementación desta callback arriba.

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/SitumARPlatformView.kt:
```diff
@@ -65,18 +68,30 @@ class SitumARPlatformView(
+            config.planeFindingMode = Config.PlaneFindingMode.DISABLED
+            config.updateMode = Config.UpdateMode.LATEST_CAMERA_IMAGE
         }
+//        sceneView.sessionConfiguration = { session, config ->
```
Non queremos código comentado!
Manías de móbil xDD

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/scene/ARQuality.kt:
```diff
@@ -0,0 +1,287 @@
+        arLocationBuffer: MutableList<LocationCoordinates>,
+        situmLocationBuffer: MutableList<LocationCoordinates>
+    ): Double {
+        var transformedARTrajectory = transformTrajectory(arLocationBuffer)
```
Revisa o lint porfa, parece que hai algúns warnings. Este por exemplo é leve (usar val en vez de var), pero hai máis por aí abaixo.

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/scene/ARQuality.kt:
```diff
@@ -0,0 +1,287 @@
+        if (currentRefreshThreshold.value > 0.20 &&
+            currentTimestamp - currentRefreshThreshold.timestamp > 1000
+        ) {
+            currentRefreshThreshold.value =
```
Aquí podes usar "-=" e aforras liñas.

Creo que nesta clase casi todos os métodos poden ser privados. É algo leve, pero o lint protesta e resulta molesto.

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/scene/DebugInfo.kt:
```diff
@@ -0,0 +1,155 @@
+        }
+        this.addView(buttonContainer)
+
+// Crear el botón grande en la esquina superior derecha
```
Poderías revisar a indentación porfa?

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/scene/DebugInfo.kt:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    fun updateText(text: String) {
+        textView.setText(text)
```
Chorradilla: en kotlin podes (e debes) facer `textView.text = text`.

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/scene/LocationCoordinates.kt:
```diff
@@ -0,0 +1,71 @@
+import kotlin.math.sqrt
+
+
+data class LocationCoordinates(
```
Esta clase pode interesarche ou non, según prefiras, pero déixoche caer:
Si o único que necesitas é utilizar novos métodos que a nosa Location non ten, podes definilos como extensións e aforras unha clase enteira, constructores engorrosos, etc:
https://kotlinlang.org/docs/extensions.html

Si ademais queres que funcione como "adapter" para separarte das Location de Situm SDK, daquela perfecto como está.

R-lagoman replied:
@germanRodriguezGarcia tiveches en conta este comentario?

germanRodriguezGarcia replied:
Perdoa, que se me pasara responder a esto. Non controlaba o das extensións.

En principio pensaba deixalo así porque utilizo esto tanto para evaluar Situm Locations como Locations de AR. Entendo que para as de AR podería utilizar Locations tamen pero habería campos sen sentido ou con valores incoherentes.

R-lagoman replied:
Ok! Comentábao por si che servía de algo. Si hai motivos para ter esta clase, adiante 👍

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/scene/LocationCoordinates.kt:
```diff
@@ -0,0 +1,71 @@
+        )
+    }
+
+    fun angularDistanceTo(other: LocationCoordinates): Double {
```
Revisa porfa, creo que non se usan todos os métodos desta clase.

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/scene/ModelUtils.kt:
```diff
@@ -0,0 +1,20 @@
+}
+
+
+class ModelUtils {
```
Esto úsase? Si non se usa, o nome do ficheiro debería cambiar?

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/scene/PoiUtils.kt:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+
+    fun calculateDistance(location1: Location, point: Point): Double {
```
Novamente revisar porfa, creo que non se usa todo nesta clase.

germanRodriguezGarcia committed:
pr changes

germanRodriguezGarcia committed:
Changed geofence mcustomfield models format. When loading models, try to load from local by model name, if not present, try to download from url field.
Fixed model positioning function to put models in the current view cone of the  camera

germanRodriguezGarcia committed:
refactor and check if new model position is already occupied

germanRodriguezGarcia committed:
check periodically if there are models nearby, if not redraw them

germanRodriguezGarcia committed:
disable shrink resources in order to keep models under raw res dir

germanRodriguezGarcia committed:
Extract constants. Format Code

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/ARController.kt:
```diff
@@ -12,33 +20,83 @@ class ARController(
         const val TAG = "Situm> AR>"
     }
 
+//    init {
```
O código comentado a futuro crearía dudas, nunca deixamos. De feito incluso o lint que trae Android Studio ten unha regla para eso xD

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/scene/ARSceneHandler.kt:
```diff
@@ -88,84 +227,824 @@ class ARSceneHandler(
+
+    fun setDebugMode(debug: Boolean) {
+        onDebug = debug;
+    }
```
Esta clase é un pouco peligro, pode ser? Ten xa máis de 1K liñas. Eu intentaría buscar a forma de estructuralo en clases máis pequenas con responsabilidades limitadas (xa sabes, principio de mínima responsabilidade). Non bloquearei a PR por esto, pero é o que eu faría.

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/scene/Utils.kt:
```diff
@@ -0,0 +1,189 @@
+import kotlin.system.measureTimeMillis
+
+val TAG = "Situm> AR>"
+data class RelativePosition(
```
Esta clase úsase?

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/scene/Utils.kt:
```diff
@@ -0,0 +1,189 @@
+import kotlin.random.Random
+import kotlin.system.measureTimeMillis
+
+val TAG = "Situm> AR>"
```
Este TAG penso que pode ser unha constante en vez dun val.

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/scene/Utils.kt:
```diff
@@ -0,0 +1,189 @@
+}
+
+fun calculateDistance(start: Vector3, end: Vector3): Float {
+    return Math.sqrt(
```
En kotlin parece que tes funcións específicas, así evitas o uso dos estáticos.

R-lagoman commented on android/src/main/kotlin/com/situm/flutter/ar/situm_ar/scene/Utils.kt:
```diff
@@ -0,0 +1,189 @@
+}
+
+
+fun getCameraDirection(worldRotation: Rotation): Rotation {
```
Podes revisar o lint? Hai varios métodos que non se usan nesta clase.

germanRodriguezGarcia committed:
refactor

germanRodriguezGarcia committed:
lint fixes

germanRodriguezGarcia committed:
refactor wip

germanRodriguezGarcia committed:
fixes

germanRodriguezGarcia committed:
fixes on destination reached

germanRodriguezGarcia committed:
minor fixes

R-lagoman reviewed (approved):
Ok, sigue habendo cargas de responsabilidade en certos obxectos que eu separaría en capas (descargas, parseados, etc). Pero non vou olvidar que se trata dun prototipo, así que adiante!

germanRodriguezGarcia committed:
Merge branch 'develop-ar-native' into feature/PRO-3484-android
